SnakeGame/model.py

```diff
@@ -22,12 +10,113 @@ def __init__(self):
         self.initSnakeLen = self.DEFAULT_INIT_SNAKE_LEN
 
     def init(self):
-        # start with an empty grid
-        self.grid = [list([self.CellContent.Empty] * self.gridSize[0]) for x in range(self.gridSize[1])]
-        coord = [0,0]
-        for i in range(self.initSnakeLen - 1):
-            self.grid[coord[0]][coord[1]] = self.CellContent.SnakeBody
-            # lay out the snake horizontally at the top left
-            coord[1] += 1
-        self.grid[coord[0]][coord[1]] = self.CellContent.SnakeHead
-        self.snakeDirection = self.Direction.Right
+        gridWidth = self.gridSize[0]
+        snakeBody = []
+        # we place the snake in the top left starting by the tail
+        x = 0
+        y = 0
+        direction = Direction.Right
+        for i in range(self.initSnakeLen):
+            # if we reach the edge, we go one row down and start filling backwards
+            coord = [x, y]
+            snakeBody.append(coord)
+            if x == 0:
+                if direction == Direction.Left:
+                    direction = Direction.Down
+                else:
+                    direction = Direction.Right
+            elif x == gridWidth - 1:
+                if direction == Direction.Right:
+                    direction = Direction.Down
+                else:
+                    direction = Direction.Left
+            if direction == Direction.Right:
+                x += 1
+            elif direction == Direction.Left:
+                x -= 1
+            else:
+                y += 1
+            # note that we do not handle the erroneous case of snake length being bigger than grid size
+        snakeBody.reverse()
+        self.snake = Snake(snakeBody, direction)
+
+    def moveSnake(self, growSnake=False):
+        '''Moves the snake in the Direction set in the object.
+        If grow is True, the snake will grow by one block while moving.'''
+        gridWidth = self.gridSize[0]
+        gridLength = self.gridSize[1]
+        # we only need to determine where the head goes next
+        # and then all the other blocks switch places with their previousBlock each
+        head = self.snake.body[0]
+        previousBlock = head.copy()
```

Comment:
Unfamiliar with .copy() ... assuming method for reference in python :)

Reply:
More or less... Python treats everything as objects, so we need to make a shallow copy of the list to make the information does not get overwritten in the lines below.